pkg/cmd/pr/shared/finder.go

```diff
@@ -0,0 +1,333 @@
+package shared
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/api"
+	"github.com/cli/cli/context"
+	"github.com/cli/cli/git"
+	"github.com/cli/cli/internal/ghrepo"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/set"
+)
+
+type PRFinder interface {
+	Find(opts FindOptions) (*api.PullRequest, ghrepo.Interface, error)
+}
+
+type progressIndicator interface {
+	StartProgressIndicator()
+	StopProgressIndicator()
+}
+
+type finder struct {
+	baseRepoFn func() (ghrepo.Interface, error)
+	branchFn   func() (string, error)
+	remotesFn  func() (context.Remotes, error)
+	httpClient func() (*http.Client, error)
+	progress   progressIndicator
+
+	repo       ghrepo.Interface
+	prNumber   int
+	branchName string
+}
+
+func NewFinder(factory *cmdutil.Factory) PRFinder {
+	if runCommandFinder != nil {
+		f := runCommandFinder
+		runCommandFinder = &mockFinder{err: errors.New("you must use a RunCommandFinder to stub PR lookups")}
+		return f
+	}
```

Comment:
I'm not so keen on adding the PRFinder to Factory since factory is something that is shared between all commands, and only a subset of commands need to find PRs. Moreover, if we managed to get rid of `runCommand` tests, this clumsy approach wouldn't be needed at all. So I'm hoping that we can figure out a migration path away from `runCommand`-style.

> I would like to see the name change to be more explicit that it is for testing purposes

👍 Agreed!